### PR TITLE
optimize `min(pk)` and `max(pk)`

### PR DESCRIPTION
**Pull request by JCOR11599**
This PR adds an optimization to queries that have a `MIN` or `MAX` aggregation over a `PRIMARY KEY` column.
Since indexes are already sorted (and we support a reverse iterator) we can look at the first/last row to answer queries in this form.

The new analyzer rule, `replaceAgg`, converts queries of the format `select max(pk) ... from ...` to the equivalent `select pk ... from ... order by pk limit 1`. Then, we depend on an `replacePkSort` to apply `IndexedTableAccess`

Additionally, this PR has `replacePkSort` optimization apply to queries that have filters (specifically those that were pushed down to IndexedTableAccess)
There is also a some refactoring and tidying up.

Fixes https://github.com/dolthub/dolt/issues/6793

### CODE REVIEW

File: sql/analyzer/replace_pk_sort.go [hunk 1]
```diff
@@ -16,57 +20,72 @@ func replacePkSort(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.Scope,
 func replacePkSortHelper(ctx *sql.Context, scope *plan.Scope, node sql.Node, sortNode *plan.Sort) (sql.Node, transform.TreeIdentity, error) {
 	switch n := node.(type) {
 	case *plan.Sort:
-		sortNode = n // TODO: this only preserves the most recent Sort node
-	// TODO: make this work with IndexedTableAccess, if we are statically sorting by the same col
-	case *plan.ResolvedTable:
-		// No sort node above this, so do nothing
-		if sortNode == nil {
+		sortNode = n // lowest parent sort node
+	case *plan.IndexedTableAccess:
```
Comment (max-hoffman) [inline sql/analyzer/replace_pk_sort.go:24]: should we have some tests for this new behavior outside of min/max? Or does this only apply to min/max?

File: sql/analyzer/replace_pk_sort.go [hunk 2]
```diff
@@ -123,3 +142,143 @@ func replacePkSortHelper(ctx *sql.Context, scope *plan.Scope, node sql.Node, sor
...
+		}
+		// TODO: optimize when there are multiple aggregations; use LATERAL JOINS
+		if len(gb.SelectedExprs) != 1 || len(gb.GroupByExprs) != 0 {
+			return n, transform.SameTree, nil
+		}
+
+		var idxTbl sql.IndexAddressableTable
+		transform.Inspect(gb.Child, func(n sql.Node) bool {
+			switch t := gb.Child.(type) {
+			case *plan.IndexedTableAccess:
```
Comment (max-hoffman) [inline sql/analyzer/replace_pk_sort.go:166]: a LOOKUP_JOIN might be a false positive here, indexed table but values won't be in order

File: sql/analyzer/replace_pk_sort.go [hunk 3]
```diff
@@ -123,3 +142,143 @@ func replacePkSortHelper(ctx *sql.Context, scope *plan.Scope, node sql.Node, sor
...
+		// TODO: support secondary indexes
+		pkIdx, err := getPKIndex(ctx, idxTbl)
+		if err != nil {
+			return nil, transform.SameTree, err
+		}
+		if pkIdx == nil {
+			return n, transform.SameTree, nil
+		}
+		// since we're only supporting one aggregation, it must be on the first column of the primary key
+		if !strings.EqualFold(pkIdx.Expressions()[0], sf.Column.String()) {
```
Comment (max-hoffman) [inline sql/analyzer/replace_pk_sort.go:219]: i think index expressions are table-prefixed, like `table.col`. Are the sort fields also always going to be prefixed the same way? Or this there a potential gap where one is table/col and the other is col?

File: sql/analyzer/replace_pk_sort.go [hunk 4]
```diff
@@ -123,3 +142,143 @@ func replacePkSortHelper(ctx *sql.Context, scope *plan.Scope, node sql.Node, sor
...
+			return n, transform.SameTree, nil
+		}
+
+		// replace all aggs in proj.Projections with GetField
+		name := gb.SelectedExprs[0].String()
+		newProjs, _, err := transform.Exprs(proj.Projections, func(e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
```
Comment (max-hoffman) [inline sql/analyzer/replace_pk_sort.go:225]: This is a bit roundabout given the limited match pattern, I think you can just inline the single `sf.Column` as the projection. This maybe points to how the match could be more general. We could match an arbitrary 1 projection with an aggregation nested somewhere inside, like `select max(pk)+1 ....`. Not sure if we want that on the first pass.

**Review by max-hoffman — APPROVED** [review]
Good as a first pass I think, will make a lot of simple aggregations faster. Just some small edge case comments.

File: sql/analyzer/replace_pk_sort.go (hunk 4 above)
Comment (JCOR11599) [reply sql/analyzer/replace_pk_sort.go:225]: I did it this way to preserve any aliases placed over the GetFields.

File: sql/analyzer/replace_pk_sort.go (hunk 1 above)
Comment (JCOR11599) [reply sql/analyzer/replace_pk_sort.go:24]: This also impacts sorting over primary keys with a filter that has been pushed down, will add plan tests for this.

**Pull request merged**